In [2]:
import numpy as np
import math

По схеме из презентации найдем машинный эпсилон

In [3]:
e = 1 
while 1 + e/2 != 1:
    e/=2
e

2.220446049250313e-16

Теперь найдем число разрядов в мантиссе по формуле из презентации $\epsilon = b^{1-t}$

Мы знаем эпсилон и также мы знаем основание $b$. Так как мы делили наш эпсилон на два при поиске, то основание $b$ это тоже двойка в какой-то степени. Соответственно наша формула преобразовывается в $\epsilon = 2^{1-t}$

Теперь выразим $t$ и у нас получится $\log_2 \epsilon = 1 - t \Leftrightarrow  -\log_2 \frac{\epsilon}{2} = t $

In [7]:
(-1)*math.log(e/2,2)

53.0

Получаем 53 знака в мантиссе 

Теперь найдем $\epsilon_{min}$ по формуле из презентации $ 0 < X_{0} = 2^{\epsilon_{min} - 1}$

In [18]:
e_min = -1
while 2**(e_min-1) != 0:
    e_min-=1
e_min

-1075

Но 1075 это не степень двойки, следовательно берём ближайшую степень двойки, то есть -1024. Но надо ещё вычесть единичку и получится -1023

In [16]:
e_max = 1
while 2**(float(e_max)):
    e_max += 1

OverflowError: (34, 'Result too large')

In [17]:
e_max

1024

Теперь приступим к сравнению некоторых величин

In [20]:
1 < 1 + e/2

False

На самом деле эти два числа будут равны, так как эпсилон пополам при сложении с единицей даст единицу

In [12]:
1 + e/2 < 1 + e

True

Это правда, так как при сложении с единицей просто эпсилон даст очень малое число, а эпсилон пополам уже нет

In [14]:
1 + e + e/2 > 1 + e

True

Левое число будет больше, так как мы сначала прибавили эпсилон и промежуточное число стало больше единицы, соответсвенно добавляя эпсилон пополам мы тоже добавим маленькое число. Также можно посмотреть на равенство, которое я напишу ниже

In [15]:
1 + e + e/2 == 1 + 1.5*e

True

In [14]:
1 + e + e/2 < 1 + e/2

False

Левое очевидно больше, так как правое равно единице, а про левое я расписал ячейкой выше

In [16]:
1 < 1 + e + e/2

True

Схожий случай с предыдущим

In [21]:
1 < 1 + e

True

Единица плюс эпсилон больше единицы по определению машинного эпсилон



Теперь перейдем к следующим сравнениям

In [33]:
(1 + 10**(-16)) + 10**(-16) < 1 + (10**(-16) + 10**(-16))

True

Это правда, так как поочередно прибавляя к единице очень маленькое число (как это происходит слева) у нас остается все равно единица, в отличие от правого числа, где мы сначала сложили два малых числа